gdrive/analytics_client.py

```diff
@@ -0,0 +1,109 @@
+import datetime
+
+from google.oauth2 import service_account
+from google.analytics.admin import AnalyticsAdminServiceClient
+from google.analytics.data_v1beta import BetaAnalyticsDataClient
+from google.analytics.data_v1beta.types import (
+    DateRange,
+    Dimension,
+    Metric,
+    RunReportRequest,
+)
+
+import logging
+import pandas as pd
+
+from gdrive import settings
+
+log = logging.getLogger(__name__)
+
+creds = service_account.Credentials.from_service_account_info(settings.CREDENTIALS)
+
+"""
+Client for the Google Analytics (GA4) API
+
+This class contains functions relating to downloading analytics data
+for the IDVA flow.
+"""
+
+
+def download(
+    property_id, target_date: datetime, end_date: datetime = None
+) -> pd.DataFrame:
+    """
+    Access Google Analytics (GA4) api and download desired analytics report.
+    """
+    if end_date is None:
+        end_date = target_date
+
+    request = RunReportRequest(
+        property=f"properties/{property_id}",
+        limit="250",
+        # https://developers.google.com/analytics/devguides/reporting/data/v1/api-schema
+        dimensions=[
+            Dimension(name="eventName"),
+            Dimension(name="firstUserCampaignName"),
+            Dimension(name="firstUserMedium"),
+            Dimension(name="firstUserSource"),
+            Dimension(name="isConversionEvent"),
+            Dimension(name="linkUrl"),
+        ],
+        metrics=[
+            Metric(name="eventCount"),
+            Metric(name="sessions"),
+            Metric(name="totalUsers"),
+            Metric(name="eventCountPerUser"),
+            Metric(name="conversions"),
+        ],
+        date_ranges=[
+            DateRange(
+                start_date=format_date_for_api(target_date),
+                end_date=format_date_for_api(end_date),
+            )
+        ],
+    )
+
+    return create_df_from_analytics_response(
```

Comment:
I think this conversion should be done outside this function to more clearly indicate the order of operations